edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs

```diff
@@ -157,16 +177,27 @@ async void ClientCloudConnectionEstablished(object sender, IIdentity identity)
 
         async Task ProcessExistingSubscriptions(string id)
         {
-            Option<ICloudProxy> cloudProxy = await this.ConnectionManager.GetCloudConnection(id);
-            Option<IReadOnlyDictionary<DeviceSubscription, bool>> subscriptions = this.ConnectionManager.GetSubscriptions(id);
-            await subscriptions.ForEachAsync(
-                async s =>
-                {
-                    foreach (KeyValuePair<DeviceSubscription, bool> subscription in s)
+            // Set a flag for an identity that temporarily disables subscription processing for other threads while we process
+            // the identity's subscriptions, so we don't trigger processing subscriptions multiple times for the same identity concurrently.
+            if (!this.subscriptionsBeingProcessed.TryAdd(id, true))
+            {
+                // Identity subscription already being processed. Skip it.
+                Events.SkippingProcessingSubscription(id);
+            }
+            else
+            {
+                Option<ICloudProxy> cloudProxy = await this.ConnectionManager.GetCloudConnection(id);
+                Option<IReadOnlyDictionary<DeviceSubscription, bool>> subscriptions = this.ConnectionManager.GetSubscriptions(id);
+                await subscriptions.ForEachAsync(
+                    async s =>
                     {
-                        await this.ProcessSubscriptionWithRetry(id, cloudProxy, subscription.Key, subscription.Value);
-                    }
-                });
+                        foreach (KeyValuePair<DeviceSubscription, bool> subscription in s)
+                        {
+                            await this.ProcessSubscriptionWithRetry(id, cloudProxy, subscription.Key, subscription.Value);
+                        }
+                    });
+                this.subscriptionsBeingProcessed.TryRemove(id, out _);
```

Comment:
should have try finally and remove in finally to make sure it gets removed after it was processed

Reply:
+1